Avoid terminating service worker processes for deleting website data

szewai opened the pull request:
#### 4ee566546310303a1411e417006e9cb06a43045a
<pre>
Avoid terminating service worker processes for deleting website data
<a href="https://rdar.apple.com/151327930">rdar://151327930</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=293890">https://bugs.webkit.org/show_bug.cgi?id=293890</a>

Reviewed by Chris Dumez.

In 290386@main, we started to terminate service worker process before deleting website data, to ensure service worker
process cannot create more website data after deletion starts. The problem is in existing implementation, network
process may notice service worker process is gone before it receives website data deletion request, in which case it
will ask UI process to relaunch the service worker process (when there is active service worker client). The new service
worker process might be discarded soon after the website data deletion has finished, which makes the process launch
wasteful. To fix this, now UI process does not directly terminate service worker process. Instead, network process will
wait for service worker clients to exit before starting to delete website data.

This patch also contains a drive-by fix to make m_webProcessConnectionCloseHandlers keep track of multiple handlers
for each connection. Otherwise, if two delete data requests are received, only one of them will wait for connection to
close (the other will start directly since the handler is not saved).

* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::unregisterServiceWorkerClientInternal):
(WebCore::SWServer::addHandlerIfHasControlledClients):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::removeNetworkConnectionToWebProcess):
(WebKit::NetworkProcess::performDeleteWebsiteDataTask):
(WebKit::NetworkProcess::deleteWebsiteData):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::activeWebProcesses const):
(WebKit::WebsiteDataStore::removeDataInNetworkProcess):
(WebKit::WebsiteDataStore::removeData):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/295728@main">https://commits.webkit.org/295728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36a2720c083eff70f704b91f9769233f5d5758ae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105932 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25683 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16077 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111130 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56530 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34186 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80467 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108938 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20745 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95600 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60787 "Passed tests") | 
| | [💥 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20373 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13698 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55968 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90129 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113982 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33072 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89545 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33436 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91829 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89224 "Found 100 new API test failures: /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /TestWebKit:WebKit.AboutBlankLoad, /TestWebKit:WebKit.PreventEmptyUserAgent, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/pointer-lock-permission-request, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit.Find, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebProcessExtension/dom-input-element-is-user-edited ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22749 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34087 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11903 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28610 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32997 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38408 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32743 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36092 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34341 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1688,18 +1690,32 @@ void NetworkProcess::deleteWebsiteData(PAL::SessionID sessionID, OptionSet<Websi
     // Schedule a timer in case web processes do not exit on time.
     RunLoop::currentSingleton().dispatchAfter(3_s, [protectedThis = Ref { *this }, taskIdentifier]() mutable {
-        protectedThis->performDeleteWebsiteDataTask(taskIdentifier);
+        protectedThis->performDeleteWebsiteDataTask(taskIdentifier, true);
```
Second parameter here is not readable. We should use an enum class.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1688,18 +1690,32 @@ void NetworkProcess::deleteWebsiteData(PAL::SessionID sessionID, OptionSet<Websi
+    if (websiteDataTypes.contains(WebsiteDataType::ServiceWorkerRegistrations)) {
+        CheckedPtr session = networkSession(sessionID);
+        RefPtr swServer = session ? session->swServer() : nullptr;
+        if (swServer && swServer->addIdleHandler([deleteTaskAggregator] { }))
```
The meaning of the `addIdleHandler()` return value is really not obvious. We should refactor to make it clearer.

cdumez commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -1906,6 +1910,17 @@ Ref<BackgroundFetchEngine> SWServer::protectedBackgroundFetchEngine()
     return backgroundFetchEngine();
 }
 
+bool SWServer::addIdleHandler(CompletionHandler<void()>&& completionHandler)
```
The meaning of "Idle" is a bit unclear here, I think we should rename the function to indicate what you mean by idle.

szewai committed:
Avoid terminating service worker processes for deleting website data
rdar://151327930
https://bugs.webkit.org/show_bug.cgi?id=293890

Reviewed by Chris Dumez.

In 290386@main, we started to terminate service worker process before deleting website data, to ensure service worker
process cannot create more website data after deletion starts. The problem is in existing implementation, network
process may notice service worker process is gone before it receives website data deletion request, in which case it
will ask UI process to relaunch the service worker process (when there is active service worker client). The new service
worker process might be discarded soon after the website data deletion has finished, which makes the process launch
wasteful. To fix this, now UI process does not directly terminate service worker process. Instead, network process will
wait for service worker clients to exit before starting to delete website data.

This patch also contains a drive-by fix to make m_webProcessConnectionCloseHandlers keep track of multiple handlers
for each connection. Otherwise, if two delete data requests are received, only one of them will wait for connection to
close (the other will start directly since the handler is not saved).

* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::unregisterServiceWorkerClientInternal):
(WebCore::SWServer::addHandlerIfHasControlledClients):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::removeNetworkConnectionToWebProcess):
(WebKit::NetworkProcess::performDeleteWebsiteDataTask):
(WebKit::NetworkProcess::deleteWebsiteData):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::activeWebProcesses const):
(WebKit::WebsiteDataStore::removeDataInNetworkProcess):
(WebKit::WebsiteDataStore::removeData):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/295728@main
